src/main/java/com/nvidia/spark/rapids/jni/BloomFilter.java

```diff
@@ -0,0 +1,164 @@
+  private final int numHashes;
+  private final long bloomFilterBits;
+  private DeviceMemoryBuffer bloomFilter;
+
+  static {
+    NativeDepsLoader.loadNativeDeps();
+  }
+
+  /**
+   * Construct an empty BloomFilter which uses the specified number of hashes and bits.
+   * @param numHashes The number of hashes to use when adding values and probing.
+   * @param bloomFilterBits The size of the bloom filter in bits.
+   */
+  public BloomFilter(int numHashes, long bloomFilterBits){
+    if(numHashes <= 0){
+      throw new IllegalArgumentException("Bloom filters must have a positive hash count");
+    }
+    if(bloomFilterBits <= 0){
+      throw new IllegalArgumentException("Bloom filters must have a positive number of bits");
+    }
+
+    this.numHashes = numHashes;    
+    this.bloomFilterBits = bloomFilterBits;
+    
+    long bloomFilterBytes = bloomFilterByteSize(bloomFilterBits);
+    bloomFilter = DeviceMemoryBuffer.allocate(bloomFilterBytes);
+    Cuda.asyncMemset(bloomFilter.getAddress(), (byte)0, bloomFilterBytes);
+  }  
+
+  /**
+   * Construct a BloomFilter from a pre-existing buffer, using the specified number of
+   * hashes and bits.
+   * @param numHashes The number of hashes to use when adding values and probing.
+   * @param bloomFilterBits The size of the bloom filter in bits.
+   * @param buffer The pre-existing buffer of.
+   */
+  public BloomFilter(int numHashes, long bloomFilterBits, DeviceMemoryBuffer buffer){
+    if(numHashes <= 0){
+      throw new IllegalArgumentException("Bloom filters must have a positive hash count");
+    }
+    if(bloomFilterBits <= 0){
+      throw new IllegalArgumentException("Bloom filters must have a positive number of bits");
+    }
+    if(buffer.getLength() != bloomFilterByteSize(bloomFilterBits)){
+      throw new IllegalArgumentException("Invalid pre-existing buffer passed. Size mismatch");
+    }
+
+    this.numHashes = numHashes;
+    this.bloomFilterBits = bloomFilterBits;
+    bloomFilter = buffer;
+  }
+
+  @Override
+  public void close() {
+    bloomFilter.close();
+  }
+
+  /**
+   * Insert a column of longs into the bloom filter.
+   * @param cv The column containing the values to add.
+   */
+  public void put(ColumnView cv){
+    put(bloomFilter.getAddress(), bloomFilter.getLength(), bloomFilterBits, cv.getNativeView(), numHashes);
+  }
+
+  /**
+   * Probe the bloom filter with a column of longs. Returns a column of booleans. For 
+   * each row in the output; a value of true indicates that the corresponding input value
+   * -may- be in the set of values used to build the bloom filter; a value of false indicates
+   * that the corresponding input value is conclusively not in the set of values used to build
+   * the bloom filter.
+   * @param cv The column containing the values to check.
+   * @return A boolean column indicating the results of the probe.
+   */
+  public ColumnVector probe(ColumnView cv){
```

Comment:
It would be nice to have a static method form for this.  Currently the way this is working out in the RAPIDS Accelerator, it creates one of these objects just to call this method (in the filtering expression)  then throws the object away (because it handles making the buffer spillable outside of this object).

We may end up having something similar on the filter builder side as well, where we need to make the buffer spillable between put calls, which means this object only lives for just one method call.  At that point, we should just have this be one-shot static methods rather than a state-tracking class.  It looks like we want the functionality exposed as static methods.  We can keep the object around for now until the Spark-side gels a bit further.

Reply:
Done